README
WWW-Scraper-ISBN-GoogleBooks_Driver
===================================

This module is a Perl extension for ISBN search collection that is provide by
WWW-Scraper-ISBN.

This driver searches for book information from the GoogleBooks online catalog.

INSTALLATION

To install this module type the following:

   perl Makefile.PL
   make
   make test
   make install

DEPENDENCIES

This module requires these modules:

  JSON
  WWW::Scraper::ISBN
  WWW::Scraper::ISBN::Record
  WWW::Scraper::ISBN::Driver
  WWW::Mechanize

For testing the module requires these modules:

  IO::File
  Test::More

For optional testing the module would like these modules:

  Test::CPAN::Meta
  Test::CPAN::Meta::JSON
  Test::Pod
  Test::Pod::Coverage

COPYRIGHT AND LICENSE

  Copyright (C) 2010-2014 Barbie for Miss Barbell Productions

  This distribution is free software; you can redistribute it and/or
  modify it under the Artistic Licence v2.


Changes
Revision history for WWW-Scraper-ISBN-GoogleBooks_Driver
========================================================

0.29    2014-12-18
        - fixed test data.

0.28    2014-11-27
        - fixed missing prerequisite.

0.27    2014-11-13
        - fixed license fields in META.json to be lists.
        - ping fix for Cygwin (thanks to Alexandr Ciornii).
        - extended META test suite.

0.26    2014-08-19
        - further patterns for Spanish pages.

0.25    2014-08-03
        - rework of html entity encodings.
        - re-engineered domain used to allow testing of different Google 
          websites.

0.24    2014-07-02
        - changed pattern matching to allow lists of patterns.

0.23    2014-06-19
        - added support for Czech Republic pages.
        - further patterns for Russian pages.

0.22    2014-06-16
        - fixed distribution name in META.
        - upgrade to WWW-Scraper-ISBN-1.00.
        - added ISBN validation checks.
        - reworked test failure checks.

0.21    2014-05-03
        - fixes for Moldovian and Russian pages, broken in last release.

0.20    2014-05-01
        - added support for Russian pages.

0.19    2014-04-29
        - added git repository links to metadata.
        - added support for Moldovian pages.

0.18    2013-07-07
        - added support for Austrian pages.

0.17    2013-06-23
        - added support for Finnish & Dutch pages.

0.16    2013-05-19
        - added isbn-check.pl to examples.
        - META updates.
        - updated publisher & pubdate patterns due to website changes.

0.15    2013-03-03
        - added support for French pages.

0.14    2013-02-17
        - 'Language not currently supported' handled better.
        - default lanmguage for .ch is German.
        - alternative expressions for German language patterns.
        - META resources updates.

0.13    2012-12-18
        - added Language not supported error.
        - added support for Hebrew pages.

0.12    2012-12-16
        - updated link for checking language.
        - description update, failing test fixed.
        - change file dates changed to meet W3CDTF standards.
        - updated test suite.

0.11    2012-06-10
        - added debugging to understand differences between other local sites.

0.10    2012-06-05
        - .com (and possibly others) don't always use Google's own CDN.

0.09    2012-05-27
        - GoogleBooks redirects to local sites, which do not necessarily have
          the same data for testing, or the same words for pattern matching.
          As such, language specific patterns are now included (de & en).
          If this distro does not work within your country, please let me know
          the test results, and I'll try and adapt the code. Patches are also 
          welcome too ;)

0.08    2012-05-22
        - further update to pattern matches.

0.07    2012-05-13
        - updated pattern matching due to page layout changes.
        - spelling fixes. (thanks to Florian Schlichting & Gregor Herrmann).
        - META tests no longer require specific versions.
        - added minimum perl version (5.006).
        - reworked Makefile.PL for clarity.

0.06    2011-05-06
        - revised link tests to use regexes.

0.05    2011-05-04
        - upgraded to Test::CPAN::Meta::JSON for tests.
        - Google now produces broken XHTML :(

0.04    2010-12-09
        - amended pingtest() in network tests.
        - added further skip test if test books are withdrawn from the site.
        - updated published date in tests.

0.03    2010-10-06
        - fixed skip count.
        - provided alternate ping test methods.

0.02    2010-09-15
        - reworked test structure, so easier to add sample sets.

0.01    2010-09-06
        - initial release


Makefile.PL
#!/usr/bin/perl

use strict;
use warnings;
use 5.006;

use ExtUtils::MakeMaker;

WriteMakefile(
    AUTHOR              => 'Barbie <barbie@cpan.org>',
    NAME                => 'WWW::Scraper::ISBN::GoogleBooks_Driver',
    VERSION_FROM        => 'lib/WWW/Scraper/ISBN/GoogleBooks_Driver.pm',
    ABSTRACT            => 'Search driver for GoogleBooks online book catalog',
    NO_META             => 1,
    PREREQ_PM           => {

        # prereqs
		'JSON'			                => '0',
        'WWW::Mechanize'                => '1.60',
        'WWW::Scraper::ISBN'            => '1.00',
        'WWW::Scraper::ISBN::Driver'    => '1.00',

        # build/test prereqs
        'Data::Dumper'                  => '0',
        'IO::File'                      => '0',
        'Test::More'                    => '0.70'

    }
);


lib/WWW/Scraper/ISBN/GoogleBooks_Driver.pm
package WWW::Scraper::ISBN::GoogleBooks_Driver;

use strict;
use warnings;
use utf8;

use vars qw($VERSION @ISA);
$VERSION = '0.29';

#--------------------------------------------------------------------------

=head1 NAME

WWW::Scraper::ISBN::GoogleBooks_Driver - Search driver for Google Books online book catalog.

=head1 SYNOPSIS

See parent class documentation (L<WWW::Scraper::ISBN::Driver>)

=head1 DESCRIPTION

Searches for book information from Google Books online book catalog

=cut

#--------------------------------------------------------------------------

###########################################################################
# Inheritence

use base qw(WWW::Scraper::ISBN::Driver);

###########################################################################
# Modules

use HTML::Entities;
use JSON;
use WWW::Mechanize;

###########################################################################
# Constants & Variables

my $DOMAIN = 'http://books.google.com';

use constant	SEARCH	=> '/books?jscmd=viewapi&callback=bookdata&bibkeys=ISBN:';
use constant	LB2G    => 453.59237;   # number of grams in a pound (lb)
use constant	OZ2G    => 28.3495231;  # number of grams in an ounce (oz)
use constant	IN2MM   => 25.4;        # number of inches in a millimetre (mm)

my %LANG = (
    'cz' => { Publisher => 'Vydavatel',     Author => 'Autor',          Title => 'Titul',   Length => [ 'Délka', qr/\QD\x{e9}lka\E/, 'D&eacute;lka' ],
                                                                                                                        Pages => [ 'Počet stran:', qr/\QPo\x{10d}et stran:\E/, 'Po&#x10D;et stran:' ] },
    'de' => { Publisher => 'Verlag',        Author => 'Autor',          Title => 'Titel',   Length => qr{L.+nge},       Pages => 'Seiten' },
    'en' => { Publisher => 'Publisher',     Author => 'Author',         Title => 'Title',   Length => 'Length',         Pages => 'pages'  },
    'es' => { Publisher => 'Editor',        Author => 'Autor',          Title => 'Título',  Length => [ 'N.º de páginas', 'N.&ordm; de p&aacute;ginas' ], 
                                                                                                                        Pages => [ 'páginas', 'p&aacute;ginas' ]  },
    'fr' => { Publisher => '.+diteur',      Author => 'Auteur',         Title => 'Titre',   Length => 'Longueur',       Pages => 'pages'  },
    'fi' => { Publisher => 'Kustantaja',    Author => 'Kirjoittaja',    Title => 'Otsikko', Length => 'Pituus',         Pages => 'sivua'  },
    'nl' => { Publisher => 'Uitgever',      Author => 'Auteur',         Title => 'Titel',   Length => 'Lengte',         Pages => [ q{pagina's}, 'pagina&#39;s' ] },
    'md' => { Publisher => 'Editor',        Author => 'Autor',          Title => 'Titlu',   Length => 'Lungime',        Pages => 'pagini'  },
    'ru' => { Publisher => ['Издатель', qr/\Q\x{418}\x{437}\x{434}\x{430}\x{442}\x{435}\x{43b}\x{44c}\E/, '&#x418;&#x437;&#x434;&#x430;&#x442;&#x435;&#x43B;&#x44C;', '&ETH;&#152;&ETH;&middot;&ETH;&acute;&ETH;&deg;&Ntilde;&#130;&ETH;&micro;&ETH;&raquo;&Ntilde;&#140;' ],
                                            Author => 'Автор',          Title => 'Название',
                                                                                            Length => [ 'Количество страниц', qr/\Q\x{41a}\x{43e}\x{43b}\x{438}\x{447}\x{435}\x{441}\x{442}\x{432}\x{43e} \x{441}\x{442}\x{440}\x{430}\x{43d}\x{438}\x{446}/, '&#x41A;&#x43E;&#x43B;&#x438;&#x447;&#x435;&#x441;&#x442;&#x432;&#x43E; &#x441;&#x442;&#x440;&#x430;&#x43D;&#x438;&#x446;', '&ETH;&#154;&ETH;&frac34;&ETH;&raquo;&ETH;&cedil;&Ntilde;&#135;&ETH;&micro;&Ntilde;&#129;&Ntilde;&#130;&ETH;&sup2;&ETH;&frac34; &Ntilde;&#129;&Ntilde;&#130;&Ntilde;&#128;&ETH;&deg;&ETH;&frac12;&ETH;&cedil;&Ntilde;&#134;' ],
                                                                                                                        Pages => [ 'Всего страниц:', qr/\Q\x{412}\x{441}\x{435}\x{433}\x{43e} \x{441}\x{442}\x{440}\x{430}\x{43d}\x{438}\x{446}:/, '&#x412;&#x441;&#x435;&#x433;&#x43E; &#x441;&#x442;&#x440;&#x430;&#x43D;&#x438;&#x446;:', '&ETH;&#146;&Ntilde;&#129;&ETH;&micro;&ETH;&sup3;&ETH;&frac34; &Ntilde;&#129;&Ntilde;&#130;&Ntilde;&#128;&ETH;&deg;&ETH;&frac12;&ETH;&cedil;&Ntilde;&#134;', '&ETH;&#146;&Ntilde;&#129;&ETH;&micro;&ETH;&sup3;&ETH;&frac34; &Ntilde;&#129;&Ntilde;&#130;&Ntilde;&#128;&ETH;&deg;&ETH;&frac12;&ETH;&cedil;&Ntilde;&#134;:' ] },
    'iw' => { Publisher => [ '\x{5d4}\x{5d5}\x{5e6}\x{5d0}\x{5d4}', '&#x5D4;&#x5D5;&#x5E6;&#x5D0;&#x5D4;' ],
                                            Author => 'Author',         Title => 'Title',   Length => [ qr/\Q\x{5d0}\x{5d5}\x{5e8}\x{5da}\E/, '××•×¨×š', '\x{5d0}\x{5d5}\x{5e8}\x{5da}', '&#x5D0;&#x5D5;&#x5E8;&#x5DA;' ],
                                                                                                                        Pages => [ qr/\Q\x{5e2}\x{5de}\x{5d5}\x{5d3}\x{5d9}\x{5dd}\E/, '×¢×ž×•×“×™×', '\x{5e2}\x{5de}\x{5d5}\x{5d3}\x{5d9}\x{5dd}', '&#x5E2;&#x5DE;&#x5D5;&#x5D3;&#x5D9;&#x5DD;' ]  }
);

#--------------------------------------------------------------------------

###########################################################################
# Public Interface

=head1 METHODS

=over 4

=item C<search>

Creates a query string, then passes the appropriate form fields to the
GoogleBooks server.

The returned page should be the correct catalog page for that ISBN. If not the
function returns zero and allows the next driver in the chain to have a go. If
a valid page is returned, the following fields are returned via the book hash:

  isbn          (now returns isbn13)
  isbn10
  isbn13
  ean13         (industry name)
  author
  title
  book_link
  image_link
  pubdate
  publisher
  description   (if available)
  pages         (if known)

The book_link and image_link refer back to the GoogleBooks website.

=back

=cut

sub search {
	my $self = shift;
	my $isbn = shift;
    my $data;
	$self->found(0);
	$self->book(undef);

    # validate and convert into EAN13 format
    my $ean = $self->convert_to_ean13($isbn);
    return $self->handler("Invalid ISBN specified")
        unless($ean);

	my $mech = WWW::Mechanize->new();
    $mech->agent_alias( 'Linux Mozilla' );

    my $search = ($ENV{GOOGLE_DOMAIN} || $DOMAIN) . SEARCH . $ean;
    eval { $mech->get( $search ) };
    return $self->handler("GoogleBooks website appears to be unavailable.")
	    if($@ || !$mech->success() || !$mech->content());

    my $json = $mech->content();

    return $self->handler("Failed to find that book on GoogleBooks website.")
	    if($json eq 'bookdata({});');

    $json =~ s/^bookdata\(//;
    $json =~ s/\);$//;

    my $code = decode_json($json);
#use Data::Dumper;
#print STDERR "\n# code=".Dumper($code);

    return $self->handler("Failed to find that book on GoogleBooks website.")
	    unless($code->{'ISBN:'.$ean} || $code->{'ISBN:'.$isbn});

    $data->{url}   = $code->{'ISBN:'.$ean }{info_url};
    $data->{url} ||= $code->{'ISBN:'.$isbn}{info_url};

    return $self->handler("Failed to find that book on GoogleBooks website.")
	    unless($data->{url});

    eval { $mech->get( $data->{url} ) };
    return $self->handler("GoogleBooks website appears to be unavailable.")
	    if($@ || !$mech->success() || !$mech->content());

	# The Book page
    #my $html = $mech->content();
    my $html = encode_entities($mech->content(),'^\n\x20-\x25\x27-\x7e');
    $html =~ s/\&amp;#39;/&#39;/sig;
    $html =~ s/\\x\(([a-z\d]+)\)/\&#$1;/sig;
    $html =~ s/&#55;/7/sig;

	return $self->handler("Failed to find that book on GoogleBooks website. [$isbn]")
		if($html =~ m!Sorry, we couldn't find any matches for!si);

#use Data::Dumper;
#print STDERR "\n# " . Dumper($data);
#print STDERR "\n# html=[$html]\n";

    $data->{url} = $mech->uri();
    my ($ccTLD) = $data->{url} =~ m{^http://[.\w]+\.google\.(\w\w)\b};

    my $lang = 'en';                                                                # English (default)
    $lang = 'de'    if($data->{url} =~ m{^http://[.\w]+\.google\.(de|ch|at)\b});    # German
    $lang = 'iw'    if($data->{url} =~ m{^http://[.\w]+\.google\.co\.il\b});        # Hebrew
    $lang = $ccTLD  if($LANG{$ccTLD});                                              # we have a ccTLD translation

	return $self->handler("Language '".uc $lang."'not currently supported, patches welcome.")
		if($lang =~ m!xx!);

    _match( $html, $data, $lang );

    # remove HTML tags
    for(qw(author)) {
        next unless(defined $data->{$_});
        $data->{$_} =~ s!<[^>]+>!!g;
    }

	# trim top and tail
	for(keys %$data) { next unless(defined $data->{$_});$data->{$_} =~ s/^\s+//;$data->{$_} =~ s/\s+$//; }

    # .com (and possibly others) don't always use Google's own CDN
    if($data->{image} =~ m!^/!) {
        my $domain = $mech->uri();
        $domain = s!^(http://[^/]+).*$!$1!;
        $data->{image} = $domain . $data->{image};
        $data->{thumb} = $data->{image};
    }

    my $url = $mech->uri();

	my $bk = {
		'ean13'		    => $data->{isbn13},
		'isbn13'		=> $data->{isbn13},
		'isbn10'		=> $data->{isbn10},
		'isbn'			=> $data->{isbn13},
		'author'		=> $data->{author},
		'title'			=> $data->{title},
		'book_link'		=> "$url",
		'image_link'	=> $data->{image},
		'thumb_link'	=> $data->{thumb},
		'pubdate'		=> $data->{pubdate},
		'publisher'		=> $data->{publisher},
		'description'   => $data->{description},
		'pages'		    => $data->{pages},
        'html'          => $html
	};

#use Data::Dumper;
#print STDERR "\n# book=".Dumper($bk);

    $self->book($bk);
	$self->found(1);
	return $self->book;
}

=head2 Private Methods

=over 4

=item C<_match>

Pattern matches for book page.

=back

=cut

sub _match {
    my ($html, $data, $lang) = @_;
    my ($publisher);

#print "\n# lang=$lang\n";

    # Some pages can present publisher text in multiple styles
    my @pubs = ref($LANG{$lang}->{Publisher}) eq 'ARRAY' ? @{$LANG{$lang}->{Publisher}} : ($LANG{$lang}->{Publisher});
    for my $pub (@pubs) {
        ($publisher) = $html =~ m!<td class="metadata_label">(?:<span[^>]*>)?$pub(?:</span>)?</td><td class="metadata_value">(?:<span[^>]*>)?([^<]+)(?:</span>)?</td>!si;
        last if($publisher);
    }
    if($publisher) {
        my @publist         = split(qr/\s*,\s*/,$publisher);
        $data->{publisher}  = $publist[0];
        $data->{pubdate}    = $publist[-1];
    }

    # Some pages can present length/pages text in multiple styles
    my @lengths = ref($LANG{$lang}->{Length}) eq 'ARRAY' ? @{$LANG{$lang}->{Length}} : ($LANG{$lang}->{Length});
    my @pages   = ref($LANG{$lang}->{Pages})  eq 'ARRAY' ? @{$LANG{$lang}->{Pages}}  : ($LANG{$lang}->{Pages});
    for my $length (@lengths) {
        for my $page (@pages) {
            ($data->{pages}) = $html =~ m!<td class="metadata_label">(?:<span[^>]*>)?$length(?:</span>)?</td><td class="metadata_value">(?:<span[^>]*>)?(\d+)\s*$page(?:</span>)?</td>!si;
            last    if($data->{pages});
            ($data->{pages}) = $html =~ m!<td class="metadata_label">(?:<span[^>]*>)?$length(?:</span>)?</td><td class="metadata_value">(?:<span[^>]*>)?\s*$page\s+(\d+)(?:</span>)?</td>!si;
            last    if($data->{pages});
        }
        last    if($data->{pages});
    }

    # get ISBN styles
    my ($isbns)         = $html =~ m!<td class="metadata_label">(?:<span[^>]*>)?ISBN(?:</span>)?</td><td class="metadata_value">(?:<span[^>]*>)?([^<]+)(?:</span>)?</td>!i;
    my (@isbns)         = split(qr/\s*,\s*/,$isbns);
    for my $value (@isbns) {
        $data->{isbn13} = $value    if(length $value == 13);
        $data->{isbn10} = $value    if(length $value == 10);
    }

#use Data::Dumper;
#print STDERR "\n# isbns=[$isbns]";
#print STDERR "\n# " . Dumper($data);

    # get other fields
    ($data->{image})                    = $html =~ m!<div class="bookcover"><img src="([^"]+)"[^>]+id=summary-frontcover[^>]*></div>!i;
    ($data->{image})                    = $html =~ m!<div class="bookcover"><a[^>]+><img src="([^"]+)"[^>]+id=summary-frontcover[^>]*></a></div>!i  unless($data->{image});
    ($data->{author})                   = $html =~ m!<td class="metadata_label">(?:<span[^>]*>)?$LANG{$lang}->{Author}(?:</span>)?</td><td class="metadata_value">(.*?)</td>!i;
    ($data->{author})                   = $html =~ m!<td class="metadata_value"><a class="primary" href=".*?"><span dir=ltr>([^<]+)</span></a></td>!si    unless($data->{author});
    ($data->{title})                    = $html =~ m!<td class="metadata_label">(?:<span[^>]*>)?$LANG{$lang}->{Title}(?:</span>)?</td><td class="metadata_value">(?:<span[^>]*>)?([^<]+)(?:</span>)?!i;
    ($data->{title})                    = $html =~ m!<meta name="title" content="([^>]+)"\s*/>! unless($data->{title});
    ($data->{description})              = $html =~ m!<meta name="description" content="([^>]+)"\s*/>!si;

    $data->{author} =~ s/"//g;
    $data->{thumb} = $data->{image};
}

1;

__END__

=head1 REQUIRES

Requires the following modules be installed:

L<WWW::Scraper::ISBN::Driver>,
L<WWW::Mechanize>
L<JSON>

=head1 SEE ALSO

L<WWW::Scraper::ISBN>,
L<WWW::Scraper::ISBN::Record>,
L<WWW::Scraper::ISBN::Driver>

=head1 BUGS, PATCHES & FIXES

There are no known bugs at the time of this release. However, if you spot a
bug or are experiencing difficulties that are not explained within the POD
documentation, please send an email to barbie@cpan.org or submit a bug to the
RT system (http://rt.cpan.org/Public/Dist/Display.html?Name=WWW-Scraper-ISBN-GoogleBooks_Driver).
However, it would help greatly if you are able to pinpoint problems or even
supply a patch.

Fixes are dependent upon their severity and my availability. Should a fix not
be forthcoming, please feel free to (politely) remind me.

=head1 AUTHOR

  Barbie, <barbie@cpan.org>
  Miss Barbell Productions, <http://www.missbarbell.co.uk/>

=head1 COPYRIGHT & LICENSE

  Copyright (C) 2010-2014 Barbie for Miss Barbell Productions

  This distribution is free software; you can redistribute it and/or
  modify it under the Artistic Licence v2.

=cut
